sana/mrs/json.py

```diff
@@ -1189,5 +1191,151 @@ def syc_encounters(request, patient_id, encounters=None):
     logging.info("finished sync_encounters")
     return render_json_response(response)
 
+@enable_logging
+def notification_get_bypt(request, id):
+    '''
+    Request Params:
+        username
+            valid username
+        password
+            valid password
+    '''
+    logging.info("entering get notification by patient procedure")
+    try:
+        notification = Notification.objects.get(patient_id=id)
+        logging.info("we finished getting the notification")
+        response = {'status': 'SUCCESS',
+                'data': cjson.decode(notification.to_json()),
+            }
+    except ObjectDoesNotExist as e:
+        et, val, tb = sys.exc_info()
+        trace = traceback.format_tb(tb)
+        error = "Exception : %s %s %s" % (et, val, trace[0])
+        for tbm in trace:
+            logging.error(tbm)
+        logging.error("Got exception while fetching notification: %s" % e)
+        response = {'status': 'FAILURE',
+            'data': "Problem while getting notification: %s" % e,
+        }
+    except MultipleObjectsReturned:
+        try:
+            notification = Notification.objects.filter(patient_id=id)
+            logging.info("we finished getting the notification")
+            response = {'status': 'SUCCESS',
+                    'data': [cjson.decode(d.to_json()) for d in notification],
```

Comment:
So there's three cases here:

| result of `Notification.objects.get(patient_id=id)` | `response['data']` |
|-----|-----|
| exactly one object | `<that object>` |
| exactly 0 objects | `"Problem while getting notification: <error>"` | 
| > 1 object | `[<object 1>, <object 2>, <etc>]` |

That puts three different datatypes into `response['data']` in each situation: in the first case, it's a JSON object, in the second it's a string and in the third it's an array. 

The way I would handle this is to instead always return an array in `response['data']`. In the first case, it can be an array with just one element. In the second case, it can be an array with no elements (you can pass the status string along in another field of `response`, like `response['message']`. Then in the third case, do as you're already doing. The advantage of that is that the client gets one consistent datatype in the `data` field - it doesn't need any conditional logic around what data structure to deserialize this to if it knows it's always going to be an array.

If you do this, you can condense this `try`/`except`/`except` into just one call to `Notification.objects.filter(patient_id=id)`. That will return the array you want to serialize, be it empty or having one object or multiple or whatever.

That also lets you get away from this structure where you have a try/except inside of a try/except which can be a little difficult to follow along with.

Hope that makes sense!